definition_validator/validator.cc

```diff
@@ -189,6 +189,42 @@ void validateOverriding(const core::GlobalState &gs, core::SymbolRef method) {
     }
 }
 
+void validateFinal(const core::GlobalState &gs, core::SymbolRef klass, optional<core::SymbolRef> nonSingleton) {
```

Comment:
Instead of taking this as a parameter, can you look it up on the symbol you're passed?

Given a singleton class, `lookupAttachedClass` will go back to the non-singleton class.

Reply:
e.g., you should be able to replace this `nonSingleton.has_value()` with `klass->data(ctx)->lookupSingletonClass(ctx).exists()`